### PR TITLE
8285521: Minor improvements in java.net.URI

### PR DESCRIPTION
**Pull request by stsypanov**
- use `String.equalsIgnoreCase()` instead of hand-written code relying on `String.charAt()`
- use `String.compareToIgnoreCase()` instead of hand-written code relying on `String.charAt()`
- drop branches that are never executed
- drop unused argument from `URI.resolvePath()`
- simplify String-related operations

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8285521](https://bugs.openjdk.org/browse/JDK-8285521): Minor improvements in java.net.URI

### Reviewers
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/8397/head:pull/8397` \
`$ git checkout pull/8397`

Update a local copy of the PR: \
`$ git checkout pull/8397` \
`$ git pull https://git.openjdk.org/jdk pull/8397/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8397`

View PR using the GUI difftool: \
`$ git pr show -t 8397`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/8397.diff">https://git.openjdk.org/jdk/pull/8397.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/java/net/URI.java [hunk 1]
```diff
@@ -1928,7 +1918,7 @@ private static int percentNormalizedComparison(String s, String t,
                 if ((sn != tn) && testForEquality)
                     return sn - tn;
                 int val = 0;
-                int n = sn < tn ? sn : tn;
+                int n = Math.min(sn, tn);
```
Comment (dfuch) [inline src/java.base/share/classes/java/net/URI.java:1921]: Can we drop this change? I wouldn't like `java.net.URI` to gratuitously trigger the loading of the Math class.
For the rest of the proposed changes, I will need to study them carefully and that may take some time.
Thanks!

Comment (cl4es) [reply src/java.base/share/classes/java/net/URI.java:1921]: Not sure it matters that much? `Math` is already unconditionally loaded on bootstrap, though slightly after `URI` in the bootstrap sequence.

Comment (stsypanov) [reply src/java.base/share/classes/java/net/URI.java:1921]: @dfuch I've looked into class loading order and I see this:
```
...
[0.179s][info][class,init] 38 Initializing 'java/util/ImmutableCollections$MapN'(no method) (0x000000080013d4b8)
[0.180s][info][class,init] 39 Initializing 'java/lang/StringLatin1' (0x0000000800021cb8)
[0.180s][info][class,init] 40 Initializing 'java/lang/Math' (0x0000000800022828)
[0.180s][info][class,init] 41 Initializing 'java/lang/Number'(no method) (0x0000000800030080)
[0.180s][info][class,init] 42 Initializing 'java/lang/Float' (0x000000080002fe10)
...
[0.234s][info][class,init] 194 Initializing 'java/lang/module/ModuleDescriptor$Requires$Modifier' (0x00000008001def98)
[0.234s][info][class,init] 195 Initializing 'java/lang/module/ModuleDescriptor$Provides'(no method) (0x00000008001dd800)
[0.235s][info][class,init] 196 Initializing 'java/net/URI' (0x0000000800142ff0)
[0.236s][info][class,init] 197 Initializing 'java/net/URI$1'(no method) (0x00000008001f4368)
[0.236s][info][class,init] 198 Initializing 'jdk/internal/module/SystemModuleFinders$2'(no method) (0x00000008001d4790)
[0.236s][info][class,init] 199 Initializing 'jdk/internal/module/SystemModuleFinders$3'(no method) (0x00000008001d50b0)
```
I.e. `Math` is loaded far before `URI`. Am I missing something?

**Comment by stsypanov** [comment]
Btw, I see that `java.net.URI` duplicates some methods from `sun.net.www.ParseUtil`. Should we consolidate them e.g. in a way of reusing functionality of `sun.net.www.ParseUtil` in `java.net.URI`?

**Comment by dfuch** [comment]
> Btw, I see that `java.net.URI` duplicates some methods from `sun.net.www.ParseUtil`. Should we consolidate them e.g. in a way of reusing functionality of `sun.net.www.ParseUtil` in `java.net.URI`?

I don't think the ratio benefit/risk is high enough.

File: src/java.base/share/classes/java/net/URI.java (hunk 1 above)
Comment (dfuch) [reply src/java.base/share/classes/java/net/URI.java:1921]: OK.

**Review by ExE-Boss — CHANGES_REQUESTED** [review]
The `String.equalsIgnoreCase(…)` and `String.compareToIgnoreCase(…)` changes are incorrect, as the `String.*IgnoreCase(…)` methods compare all **Unicode** code points case‑insensitively using **Unicode** rules for the current or specified locale, whereas the **URI** specification does case‑insensitive comparison only for characters in the **US‑ASCII** range [[RFC3986]].

--------------------------------------------------------------------------------

https://github.com/openjdk/jdk/blob/408a3a8e29006798071cd6f185e415bc2bc62282/src/java.base/share/classes/java/net/URI.java#L1825-L1830 https://github.com/openjdk/jdk/blob/408a3a8e29006798071cd6f185e415bc2bc62282/src/java.base/share/classes/java/net/URI.java#L1832-L1844

[RFC3986]: https://datatracker.ietf.org/doc/html/rfc3986

**Comment by stsypanov** [comment]
@ExE-Boss 

> String.*IgnoreCase(…) methods compare all Unicode code points case‑insensitively using Unicode rules for the current or specified locale, whereas the URI specification does case‑insensitive comparison only for characters in the US‑ASCII range

Aren't all the items of US-ASCII range belong to Unicode regardless of locale?

**Comment by ExE-Boss** [comment]
Note that if the current locale is Turkish, then `"I".equalsIgnoreCase("i")` (`"\u0049".equalsIgnoreCase("\u0069")`) returns `false`.

**Comment by stsypanov** [comment]
Ok, I'll revert these changes then.

**Comment by stsypanov** [comment]
Let's wait.

**Pull request closed without merging**